feat(ivy): implement pipes

marclaval opened the pull request:
This PR implements pipes in Ivy.

Some comments about the decisions taken here:
- `pipeBind1` to `pipeBind4` could be shortened if using the spread operator, but it transpiles into for loops
- `pipeBindV` has a different signature to match `objectLiteralV`
- a pipe can return a `WrappedValue`, which means that any binding can store such object. So when accessing a binding, it has to be unwrapped
- the global cache of pure pipes has been kept, a new global variable has been added to the instruction module (since we don't have a proper application state)
 => the signature of the `pipe` instruction has been changed.  The instruction is responsible for instanciating the pipe (or to use the cache)
 => the canonical spec has updated accordingly
- tests are directly ported from current Angular, most of them test the HTML (sorry @vicb)

marclaval committed:
feat(ivy): implement pipes

marclaval commented on packages/core/src/render3/definition.ts:
```diff
@@ -93,7 +93,7 @@ export function NgOnChangesFeature(definition: DirectiveDef<any>): void {
           simpleChanges = this[PRIVATE_PREFIX] = {};
         }
-        simpleChanges[pubKey] = new SimpleChange(this[privateMinKey], value, isFirstChange);
+        simpleChanges[minKey] = new SimpleChange(this[privateMinKey], value, isFirstChange);
```
This is how it works in current Angular

vicb replied:
If `minKey` is the minified key I think the current Angular does not work this way, i.e. `simpleChanges` is indexed by full property name.

Please add a comment in the code to explain what this code does and the intended user behavior.

marclaval replied:
Adressed in https://github.com/angular/angular/pull/22352

marclaval commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -233,7 +233,7 @@ export function destroyViewTree(rootView: LView): void {
 
     if (next == null) {
-      while (viewOrContainer && !viewOrContainer !.next) {
+      while (viewOrContainer && !viewOrContainer !.next && viewOrContainer !== rootView) {
```
if the `viewOrContainer` is the `rootView`, then the cleanup is done twice

vicb replied:
Is there a test ?
Add a comment.

marclaval replied:
Adressed in https://github.com/angular/angular/pull/22350

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1767,6 +1776,14 @@ export function getRenderer(): Renderer3 {
+}
+
+export function getPurePipeCache(): Map<PipeDef<any>, any> {
+  return purePipeCache;
```
I think it would make more sense to move the  `getGlobalX()` right after `globalX` ?

vicb commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -289,7 +289,7 @@ export const enum LifecycleStage {
  * data store a null value in tData to avoid a sparse array.
  */
-export type TData = (TNode | DirectiveDef<any>| null)[];
+export type TData = (TNode | DirectiveDef<any>| PipeDef<any>| null)[];
```
update the comment ?

vicb commented on packages/core/src/render3/pipe.ts:
```diff
@@ -15,8 +20,25 @@ import {PipeDef} from './interfaces/definition';
  */
-export function pipe<T>(index: number, pipeDef: PipeDef<T>, pipe: T): void {
-  throw new Error('TODO: implement!');
+export function pipe<T>(index: number, pipeDef: PipeDef<T>): void {
```
Please update the API doc (argument)

vicb replied:
actually @chuckjaz as implemented the former API in https://github.com/angular/angular/pull/22034/files#diff-b9d784499e486a026b7e92e7a1ee99eaR348

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -112,6 +114,11 @@ let bindingIndex: number;
 let cleanup: any[]|null;
 
+/**
+ * A cache of all the pure pipes of the application.
```
nit: Explain that we only ever need a single instance of each pure pire instead ?

(IMO `A cache of all the pure pipes of the application.` does not add a lot more to `purePipeCache`).

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -283,6 +290,7 @@ export function createLNode(
 function resetApplicationState() {
   isParent = false;
   previousOrParentNode = null !;
+  purePipeCache.clear();
```
not sure when `resetApplicationState` gets called.
ie what if we bootstrap 2 components ? could we be destroying a shared cache at this point ?
Add a comment on this function ?

vicb commented on packages/core/src/render3/pipe.ts:
```diff
@@ -29,7 +51,15 @@ export function pipe<T>(index: number, pipeDef: PipeDef<T>, pipe: T): void {
+  if (isPure(index)) {
+    const result = objectLiteral1(pipeInstance.transform.bind(pipeInstance), v1);
+    const value = result === NO_CHANGE ? peekBinding() : result;
+    bind(result);
```
`objectLiteral1 ` will be rename to `pureFunction` and will always return a value
see @kara pending PR

vicb commented on packages/core/src/render3/pipe.ts:
```diff
@@ -29,7 +51,15 @@ export function pipe<T>(index: number, pipeDef: PipeDef<T>, pipe: T): void {
+    const value = result === NO_CHANGE ? peekBinding() : result;
+    bind(result);
+    return value;
+  } else {
```
no `else` after a `return` ? (might be a personal preference)

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+    person = new Person();
+  });
+
+  it('should use the return value of the pipe', () => {
```
`should use the return value of the pipe` sounds like we are not testing the pipe ?

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+    expect(renderToHtml(Template, person)).toEqual('value one two default');
+  });
+
+  it('should associate pipes right-to-left', () => {
```
not sure what we test here ? 
Should it be a conformance test rather, ie given a template the pipes will be called in the expected order

vicb commented on packages/core/test/render3/imported_renderer2.ts:
```diff
@@ -64,3 +65,52 @@ export function getAnimationRendererFactory2(document: any): RendererFactory2 {
       fakeNgZone);
 }
+
+// TODO: code duplicated from ../linker/change_detection_integration_spec.ts, to be removed
```
What about 
`// START duplicated code ... // END duplicated code`

mhevery replied:
@marclaval I see that you did what @vicb asked you. Awesome, could you leave a note that this issue has been taken care of, it makes reading the PR easier. I usually leave 👍  or `fixed` note.

vicb commented on packages/core/test/render3/imported_renderer2.ts:
```diff
@@ -64,3 +65,52 @@ export function getAnimationRendererFactory2(document: any): RendererFactory2 {
+  }
+}
+
+export function patchLoggingRenderer2(rendererFactory: RendererFactory2, log: RenderLog) {
```
Add comments ?

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+    expect(renderToHtml(Template, person)).toEqual('value a b default 0 1 default');
+  });
+
+  it('should support calling pure pipes with different number of arguments', () => {
```
Not sure why ` pure pipes` here ?
Should we have Top Level
```
describe(pipes)
  it(...)
  it(...)
  describe(pure)
    it(...)
  describe(not pure)
    it(...)
```

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+    expect(renderLog.log).toEqual([]);
+  });
+
+  it('should unwrap the wrapped value and force a change', () => {
```
about pure/not pure would make more sense to get the diff from this test vs the previous ?

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+  });
+
+  it('should record unwrapped values via ngOnChanges', () => {
+    @Directive({selector: '[testDirective]'})
```
I very much prefer to have `Directive` / `Pipe` inside the it (aka new style testing) vs the having all the Pipe at the end of the file. Could you change that ?

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+  });
+
+  it('should call pure pipes that are used multiple times only when the arguments change', () => {
+    function Template(person: Person, cm: boolean) {
```
Not sure of the added value of this test ?
Are we testing anything different from the previous one ?

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+      }
+      elementProperty(0, 'someProp', bind(pipeBind1(1, true)));
+      elementProperty(2, 'someProp', bind(pipeBind1(3, true)));
+      pipeInstances.push(memory<CountingPipe>(1), memory(3));
```
use `<CountingPipe>` consistently ?

vicb commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,539 @@
+    expect(pipeInstances[1]).toBe(pipeInstances[0]);
+    expect(pipeInstances[2]).toBeAnInstanceOf(CountingPipe);
+    expect(pipeInstances[2]).toBe(pipeInstances[0]);
+    expect(pipeInstances[3]).toBeAnInstanceOf(CountingPipe);
```
the instanceOf[1...3] tests are redundant, can be removed

vicb commented:
Again I think we should have a clear distinction between
- unit tests,
- integration tests,
- conformance tests.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -112,6 +114,11 @@ let bindingIndex: number;
+/**
+ * A cache of all the pure pipes of the application.
+ */
+const purePipeCache: Map<PipeDef<any>, any> = new Map<PipeDef<any>, any>();
```
This is a global, which will span multiple apps. A function can be pure, but it may also inject things into its constructor, and what is visible depends on the location of the pipe. So at best we can have cache on template level, not on app level. (see later comment)

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -15,8 +20,25 @@ import {PipeDef} from './interfaces/definition';
-  throw new Error('TODO: implement!');
+export function pipe<T>(index: number, pipeDef: PipeDef<T>): void {
+  const tView = getTView();
+  tView.data[index] = pipeDef;
```
why do we need to store this? I don't think that is needed.

marclaval replied:
We need it to know if the pipe is pure

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -15,8 +20,25 @@ import {PipeDef} from './interfaces/definition';
+  let pipeInstance: T;
+  if (pipeDef.pure) {
+    const purePipeCache = getPurePipeCache();
+    if (purePipeCache.has(pipeDef)) {
```
Looking up values from cache is not the fastest. If the function is pure could we store it in the `TView`? That way our cache is per view (rather than per app). It also means that we don't have to have a global map.  I would imagine something like this:

```
if (pipeDef.pure) {
  pipeInstance = tView.data[index];
} else {
  pipeInstance = pipeDef.n();
}
```

JoostK commented on packages/core/src/render3/definition.ts:
```diff
@@ -176,5 +176,9 @@ export const defineDirective = defineComponent as<T>(directiveDefinition: Direct
\ No newline at end of file
+  return <PipeDef<any>>{
+    n: factory,
+    pure: pure !== false,
```
@mhevery Huh?? This simply states that pure defaults to true if not explicitly set to false.

trotyl replied:
`false !== false` is `false`, isn't it?

kara commented on packages/core/src/render3/pipe.ts:
```diff
@@ -29,7 +51,15 @@ export function pipe<T>(index: number, pipeDef: PipeDef<T>, pipe: T): void {
+  const pipeInstance = memory<PipeTransform>(index);
+  if (isPure(index)) {
+    const result = objectLiteral1(pipeInstance.transform.bind(pipeInstance), v1);
+    const value = result === NO_CHANGE ? peekBinding() : result;
```
The rename PR hasn't gone in yet:

https://github.com/angular/angular/pull/22214

marclaval commented:
I've updated the PR thanks to your comments. Summary of the changes:
- the lifecycle issues that were discovered have been moved to their own PRs: #22350 and #22352 
- as the pure pipe cache is per template in current Angular, the global cache has been removed. It has been agreed with @chuckjaz to have more work done by the compiler. The canonical spec has been updated accordingly.
- it has been rebased to use pure functions, which I modified as per @mhevery suggestion. I also realised that these pure functions can get `WrappedValue`, so I have done some changes to handle this case.
- tests have been reorganised and improved following @vicb comments.

chuckjaz commented on packages/core/test/render3/compiler_canonical/compiler_canonical_spec.ts:
```diff
@@ -1103,31 +1118,85 @@ describe('compiler specification', () => {
+          tag: 'my-app',
+          factory: function MyApp_Factory() { return new MyApp(); },
+          template: function MyApp_Template(ctx: $MyApp$, cm: $boolean$) {
+            let pi: MyPurePipe;
```
This should be:

```
let $pi$: $MyPurePipe$;
```

and a `type $MyPurePipe$ = MyPurePipe;` be included outside the `// NORMATIVE` block.

This makes it easy for me to cut and paste the normative block into the compiler tests and we can easily see when they are out of sync.

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1424,15 +1424,16 @@ function initBindings() {
     initBindings();
-    return data[bindingIndex++] = value;
+    data[bindingIndex++] = value;
+    return unwrap(value);
```
you should also probably update `bindingUpdated`(and the tests there)

marclaval replied:
I don't think so as we want to store the wrapped values, let's double check

mhevery replied:
I think `unwrap` should be used from pipes only. `unwrap` is not small and we should not be pulling in the code if there are no pipes.

vicb commented on packages/core/src/render3/pipe.ts:
```diff
@@ -6,17 +6,39 @@
+    (tView.destroyHooks || (tView.destroyHooks = [])).push(index, pipeDef.onDestroy);
+  }
+  let pipeInstance: T;
+  if (pipeDef.pure) {
```
looks the same as
```
const p = pipeDef.pure && firstInstance ? firstInstance : pipeDef.n();
```

marclaval replied:
👍

vicb commented on packages/core/src/render3/pipe.ts:
```diff
@@ -87,5 +117,11 @@ export function pipeBind4(index: number, v1: any, v2: any, v3: any, v4: any): an
\ No newline at end of file
+  const pipeInstance = load<PipeTransform>(index);
+  return isPure(index) ? pureFunctionV(pipeInstance.transform, values, pipeInstance) :
+                         pipeInstance.transform.apply(pipeInstance, values);
```
`pipeInstance.transform(pipeInstance, ...values);` ?

marclaval replied:
The spread operator would be transpiled into a for loop, which doesn't seem to be a good idea.

marclaval committed:
fixup! feat(ivy): implement pipes

marclaval commented:
PR updated: comments have been integrated, the disabled tests have been enabled, and one more test has been added to ensure that we always unwrap values before calling a Pipe's transform function.

kara commented on packages/core/test/render3/pipe_spec.ts:
```diff
@@ -0,0 +1,512 @@
+
+      person.age = 15;
+      renderToHtml(Template, person);
+      expect(log).toEqual(['pipeWithOnDestroy - ngOnDestroy']);
```
If you set the age to 25, then 15 again, I think this test will fail. It should be 2 logs, but it will be 3.

marclaval replied:
Done

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1424,15 +1424,16 @@ function initBindings() {
   }
   bindingIndex++;
-  return changed ? value : NO_CHANGE;
+  return changed ? unwrap(value) : NO_CHANGE;
```
remove `unwrap`

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -6,17 +6,31 @@
-  throw new Error('TODO: implement!');
+export function pipe<T>(index: number, pipeDef: PipeDef<T>, firstInstance?: T): T {
+  const tView = getTView();
+  tView.data[index] = pipeDef;
```
👍  what @kara said

marclaval replied:
👍 Done

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -87,5 +109,12 @@ export function pipeBind4(index: number, v1: any, v2: any, v3: any, v4: any): an
+  const pipeInstance = load<PipeTransform>(index);
+  return isPure(index) ?
+      pureFunctionV(pipeInstance.transform, values, pipeInstance) :
+      pipeInstance.transform.apply(pipeInstance, values.map(value => unwrap(value)));
```
do not use `map` in Ivy code as it is not as fast as a for loop.

marclaval replied:
:ok:

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -29,7 +43,9 @@ export function pipe<T>(index: number, pipeDef: PipeDef<T>, pipe: T): void {
 export function pipeBind1(index: number, v1: any): any {
-  throw new Error('TODO: implement!');
+  const pipeInstance = load<PipeTransform>(index);
+  return isPure(index) ? pureFunction1(pipeInstance.transform, v1, pipeInstance) :
```
There is an asymmetry between args in pure and non-pure case. They should be the same. Maybe have `isPure && unwrap(x)` as a way of turning pure with wrap into non-pure or something?

marclaval replied:
They can't be the same.

In the pure case, `pureFunction1` will check the bindings and then, if some input has changed, it will unwrap the values and call `transform`.
In the impure case, `transform` can be called immediately with unwrapped values.

mhevery replied:
Now that we have removed `unwrap` can we make them symmetric?

marclaval replied:
@mhevery Can you please clarify what you have in mind?

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -6,17 +6,31 @@
  * @param index Pipe index where the pipe will be stored.
  * @param pipeDef Pipe definition object for registering life cycle hooks.
- * @param pipe A Pipe instance.
+ * @param firstInstance (optional) The first instance of the pipe that can be reused for pure pipes.
```
Instead of passing `firstInstance` as a `Pipe` we could pass it as an index which would greatly simplify the work which compiler has to do. 

/cc @chuckjaz

chuckjaz replied:
Either is fine.

Storing it in a temporary that is not closure captured is simple to generate and is relatively efficient.

Passing in an index is easier. If it is already in an index then the index of the first one is an easy thing to track.

We should do whatever is consistent for the API of the instructions and the most efficient at runtime. The compiler can easily generate either.

marclaval replied:
Using index was my first idea, but it wouldn't work with embedded views, and it should.

mhevery commented on packages/core/src/render3/pipe.ts:
```diff
@@ -6,17 +6,31 @@
  * @param pipeDef Pipe definition object for registering life cycle hooks.
- * @param pipe A Pipe instance.
+ * @param firstInstance (optional) The first instance of the pipe that can be reused for pure pipes.
+ * @returns T the instance of the pipe which was stored in the bindings array.
```
I think it should just say `@returns the instance of the pipe`. Come to think of it do we even need to return anything? Can't this just be void?

marclaval replied:
Returning the instance is needed for the pure cache pipe (see other comment)

mhevery commented on packages/core/src/render3/pure_function.ts:
```diff
@@ -17,8 +17,8 @@ import {bindingUpdated, bindingUpdated2, bindingUpdated4, checkAndUpdateBinding,
-export function pureFunction0<T>(pureFn: () => T): T {
-  return getCreationMode() ? checkAndUpdateBinding(pureFn()) : consumeBinding();
+export function pureFunction0<T>(pureFn: () => T, thisArg: any = null): T {
+  return getCreationMode() ? checkAndUpdateBinding(pureFn.call(thisArg)) : consumeBinding();
```
`pureFn.call(thisArg)` will make it slow all the time. It would be much faster if you do `thisArg ? pureFn.call(thisArg) : pureFn()`

marclaval replied:
:ok:

mhevery commented on packages/core/src/render3/pure_function.ts:
```diff
@@ -184,11 +199,12 @@ export function pureFunction8(
     bindingUpdated(exps[i]) && (different = true);
   }
-  return different ? checkAndUpdateBinding(pureFn.apply(null, exps)) : consumeBinding();
+  return different ? checkAndUpdateBinding(pureFn.apply(thisArg, exps.map(exp => unwrap(exp)))) :
```
don't use `map`

marclaval replied:
:ok:

mhevery commented on packages/core/src/render3/util.ts:
```diff
@@ -16,12 +18,17 @@ export function isDifferent(a: any, b: any): boolean {
 }
 
 export function stringify(value: any): string {
+  value = unwrap(value);
```
This really bugs me that we have to do `unwrap` here. The issue is that we will pull in `unwrap` all the time even if it is not needed. We should only pull it in if we have pipes. So I think we should remove this and rely that `unwrap.toString()` will return the right value by calling `stringify` internally

marclaval replied:
:+1:

mhevery commented on packages/core/test/bundling/hello_world/bundle.golden_symbols.json:
```diff
@@ -11,6 +11,9 @@
     "name": "Symbol$1"
   },
+  {
+    "name": "WrappedValue"
```
This is wrong. Hello world should not be pulling in `WrappedValue`

marclaval commented on packages/core/src/render3/pure_function.ts:
```diff
@@ -57,10 +61,12 @@ export function pureFunction2(pureFn: (v1: any, v2: any) => any, exp1: any, exp2
-  return bindingUpdated(exp3) || different ? checkAndUpdateBinding(pureFn(exp1, exp2, exp3)) :
-                                             consumeBinding();
+  return bindingUpdated(exp3) || different ?
+      checkAndUpdateBinding(pureFn(unwrap(exp1), unwrap(exp2), unwrap(exp3))) :
```
You're right, it should, fixed

marclaval commented on packages/core/src/render3/util.ts:
```diff
@@ -16,12 +18,17 @@ export function isDifferent(a: any, b: any): boolean {
   return '' + value;
 }
 
+export function unwrap(value: any): any {
```
Added

marclaval committed:
fixup! feat(ivy): implement pipes

marclaval committed:
fixup! feat(ivy): implement pipes

marclaval commented:
As discussed with @mhevery , the support of `WrappedValue` has been removed for now. PTAL

mhevery commented:
Please resolve conflicts

kara reviewed (approved):
LGTM

marclaval commented:
Rebased

marclaval commented:
Rebased, PTAL

chuckjaz commented on packages/core/src/render3/pure_function.ts:
```diff
@@ -17,8 +17,9 @@ import {bindingUpdated, bindingUpdated2, bindingUpdated4, checkAndUpdateBinding,
  */
-export function pureFunction0<T>(pureFn: () => T): T {
-  return getCreationMode() ? checkAndUpdateBinding(pureFn()) : consumeBinding();
+export function pureFunction0<T>(pureFn: () => T, thisArg: any = null): T {
```
Consider just using `thisArg?: any` as adds `if (thisArg === void 0) { thisArg = null; }` to the beginning of the function for no good reason.

marclaval replied:
:+1:

chuckjaz commented on packages/core/test/render3/compiler_canonical/pipes_spec.ts:
```diff
@@ -52,29 +68,83 @@ xdescribe('pipes', () => {
+        tag: 'my-app',
+        factory: function MyApp_Factory() { return new MyApp(); },
+        template: function MyApp_Template(ctx: $MyApp$, cm: $boolean$) {
+          let pi: MyPurePipe;
```
Change to `let pi: $MyPurePipe$` (with the corresponding `type` declaration above) as all types are non-canonical

marclaval replied:
Done

chuckjaz commented on packages/core/test/render3/compiler_canonical/pipes_spec.ts:
```diff
@@ -52,29 +68,83 @@ xdescribe('pipes', () => {
+            $r3$.ɵT(2);
+            $r3$.ɵPp(3, $MyPurePipe_ngPipeDef$, pi);
+          }
+          $r3$.ɵt(0, $r3$.ɵi1('', $r3$.ɵpb2(1, ctx.name, ctx.size), ''));
```
Consider adding an example with an embedded template to demonstrate why the local is needed (not just a slot value) and why the `let` should be outside the block. These are both necessary but the normative example does not show an example where this is required.

marclaval replied:
Done with a structural directive

marclaval committed:
fixup! feat(ivy): implement pipes

mhevery commented:
http://test/OCL:188244226:BASE:188244256:1520465278593:6088a334
